modules/contact/src/constraints/ComputeFrictionalForceCartesianLMMechanicalContact.C

```diff
@@ -0,0 +1,358 @@
+  ADReal tangential_dof_residual;
+  ADReal tangential_dof_residual_dir;
+
+  const ADReal & tangential_vel = *_tangential_vel_ptr[0];
+
+  // Primal-dual active set strategy (PDASS)
+  if (!_has_disp_z)
+  {
+    if (normal_pressure_value.value() < _epsilon)
+      tangential_dof_residual = tangential_pressure_value;
+    else
+    {
+      const auto term_1 =
+          std::max(_mu * (normal_pressure_value + c * weighted_gap),
+                   std::abs(tangential_pressure_value + c_t * tangential_vel * _dt)) *
+          tangential_pressure_value;
+      const auto term_2 = _mu * std::max(0.0, normal_pressure_value + c * weighted_gap) *
+                          (tangential_pressure_value + c_t * tangential_vel * _dt);
+
+      tangential_dof_residual = term_1 - term_2;
+    }
+  }
+  else
+  {
+    if (normal_pressure_value.value() < _epsilon)
+    {
+      tangential_dof_residual = tangential_pressure_value;
+      tangential_dof_residual_dir = tangential_pressure_value_dir;
+    }
+    else
+    {
+      const Real epsilon_sqrt = 1.0e-48;
+
+      const auto lamdba_plus_cg = normal_pressure_value + c * weighted_gap;
+
+      std::array<ADReal, 2> lambda_t_plus_ctu;
+      lambda_t_plus_ctu[0] = tangential_pressure_value + c_t * (*_tangential_vel_ptr[0]) * _dt;
+      lambda_t_plus_ctu[1] = tangential_pressure_value_dir + c_t * (*_tangential_vel_ptr[1]) * _dt;
+
+      const auto term_1_x =
+          std::max(_mu * lamdba_plus_cg,
+                   std::sqrt(lambda_t_plus_ctu[0] * lambda_t_plus_ctu[0] +
+                             lambda_t_plus_ctu[1] * lambda_t_plus_ctu[1] + epsilon_sqrt)) *
+          tangential_pressure_value;
+
+      const auto term_1_y =
+          std::max(_mu * lamdba_plus_cg,
+                   std::sqrt(lambda_t_plus_ctu[0] * lambda_t_plus_ctu[0] +
+                             lambda_t_plus_ctu[1] * lambda_t_plus_ctu[1] + epsilon_sqrt)) *
+          tangential_pressure_value_dir;
+
+      const auto term_2_x = _mu * std::max(0.0, lamdba_plus_cg) * lambda_t_plus_ctu[0];
+
+      const auto term_2_y = _mu * std::max(0.0, lamdba_plus_cg) * lambda_t_plus_ctu[1];
+
+      tangential_dof_residual = term_1_x - term_2_x;
+      tangential_dof_residual_dir = term_1_y - term_2_y;
+    }
+  }
+
+  // Compute the friction coefficient (constant or function)
+
+  // Get index for normal constraint.
+  // We do this to get a decent Jacobian structure, which is key for the use of iterative solvers.
+  // Using old normal vector to avoid changes in the Jacobian structure within one time step
+
+  Real ny, nz;
+  // Intially, use the current normal vector
+  if (_dof_to_old_normal_vector[dof].norm() < TOLERANCE)
+  {
+    ny = _dof_to_normal_vector[dof](1);
+    nz = _dof_to_normal_vector[dof](2);
+  }
+  else
+  {
+    ny = _dof_to_old_normal_vector[dof](1);
+    nz = _dof_to_old_normal_vector[dof](2);
+  }
+  unsigned int component_normal = 0;
+  if (std::abs(ny) > 0.57735)
```

Comment:
I don't understand this number. ny could be > .57735 and nx or nz could be larger than it (this can be true up to ny = .707 = sqrt(0.5)). Where does it come from?

Reply:
A generic unit vector in 3D can be: {1/sqrt(3), 1/sqrt(3), 1/sqrt(3)}. Any deviation from this would result in a dominant component within the vector (1/sqrt(3) = 0.57735). As it is coded, it would be suboptimal for some cases in 2D, which I wasn't too concerned about honestly. The intention of this code is to avoid flagrant ill conditioning. Made some changes, let me know if you'd like something else.

Reply:
Even in a 3D simulation you could have a normal like {1/sqrt(2), 1/sqrt(2), 0}

Reply:
Yeah, I'd just go with the generic definition instead axis-aligned case. From the few simple cases I have run, this particular difference will not have a significant practical impact.

Reply:
but indeed it would be problematic if you *never* detected an "important" component, as you would not if you set the threshold to 1/sqrt(2) and then in 3D had a {1/sqrt(3), 1/sqrt(3), 1/sqrt(3)} vector. So yea I like the change you made and the variable name you gave